### ISSUE TITLE
  tf.estimator.train_and_evaluate list  of eval_spec

### ISSUE THREAD

**Issue by bhack** [tensorflow/tensorflow#16087]
Sometimes could be useful to evaluate on different clusters of the evaluation set with distinct metrics, summaries etc.
Do you plan that this interface will accept a list of `eval_spec` in the future?

**Comment by cy89** [comment]
@martinwicke @ispirmustafa any hints about future plans?

**Comment by ispirmustafa** [comment]
@xiejw could you please comment?

**Comment by xiejw** [comment]
Yes. it was planned. We are working on this.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by markpwoodward** [comment]
+1
To compare train and validation metrics "apples" to "apples", I run two evaluations, one on the validation set and one on a fixed subset of training data. To do this I call the evaluate's in a CheckpointSaverListener passed to estimator.train(...., saving_listeners), but two EvalSpec's passed to train_and_evalute(..., eval_spec) would be cleaner. Alternatively, EvalSpec could take a list of input_fn's and corresponding name's. Whichever you think is easier/better for the api.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mrezak** [comment]
Any updates on this issue? Can `tf.estimator.evaluate / tf.estimator.train_and_evaluate` get a list of `eval_spec`s yet? It'd be a very useful feature.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 76 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 93 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 108 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 123 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 138 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 153 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 168 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @xiejw: It has been 183 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by geopapa11** [comment]
> +1
> To compare train and validation metrics "apples" to "apples", I run two evaluations, one on the validation set and one on a fixed subset of training data. To do this I call the evaluate's in a CheckpointSaverListener passed to estimator.train(...., saving_listeners), but two EvalSpec's passed to train_and_evalute(..., eval_spec) would be cleaner. Alternatively, EvalSpec could take a list of input_fn's and corresponding name's. Whichever you think is easier/better for the api.

HI @markpwoodward could you please share the code to see how you are doing that? It is not very straightforward to me how we can add multiple datasets during evaluation (when we do it through TF's Estimator) and it's such a useful thing to do.

Thanks in advance!

**Comment by markpwoodward** [comment]
Here is an example of what I was talking about using estimator.train()
https://github.com/markpwoodward/active_osl/blob/master/train.py

Being able to pass a list of eval_spec's to estimator.train_and_evaluate()
would clean this up considerably, removing the need for the
CheckpointSaverListeners.

Best,
Mark

**Comment by geopapa11** [comment]
@markpwoodward thanks so much! This is immensely helpful! And yes I couldn't agree more. This added functionality in estimator.train_and_evaluate() would give so much more flexibility to test algo's performance on different datasets. Happy Holidays

**Comment by PeterMcGor** [comment]
Hi, 

just in case someone is still wondering about this, here is my workaround after some hours struggling:

In order to use `train_and_evaluate`, I include at `EvalSpec` definition a `tf.estimator.FeedFnHook` which contains a function that performs the evaluation over different datasets employing  `estimator.evaluate`. I give a different evaluation folder by the `name` parameter at `estimator.evaluate`. You will probably have to include a new `input_fn` function for `estimator.evaluate` but at least it kind of works

Hope is clearer in the following snippet:
```python
def multi_eval_hook():
  for validation_set in extra_validation_sets:
    estimator.evaluate(input_fn=lambda:super_eval_fn(val_set.get("images"), val_set.get("labels")), 
                       steps=eval_steps, name=val_set["id"])
train_spec = tf.estimator.TrainSpec(input_fn=lambda: train_fn,
                                    max_steps=train_steps * train_epochs,
                                    hooks=train_hooks)
eval_spec = tf.estimator.EvalSpec(input_fn=eval_fn,
                                  steps=eval_steps, 
                                  hooks=[tf.estimator.FeedFnHook(multi_eval_hook)])
```

**Comment by saravanabalagi** [comment]
@PeterMcGor  Is it possible to get the name we pass in to `estimator.evaluate` inside of `model_fn`? I have a custom summary writer to log stuff into that directory. Any help is appreciated.

**Comment by saravanabalagi** [comment]
I tried the workaround and I thought the evaluation was running forever. I then ran an empty multi eval hook with just logging inside.

```
def multi_eval_hook():
    logger.info("Starting additional evaluation")
    logger.info("Finished additional evaluation")

eval_spec = tf.estimator.EvalSpec(lambda: valid_input_fn(params),
                                  hooks=[tf.estimator.FeedFnHook(multi_eval_hook)], name="eval")
```
This produces 
```
2020-04-14 11:27:03.000990: I saver.py:1284] Restoring parameters from experiments/custom_loss_normalized/loss_alpha_beta=0.75_0.75/model.ckpt-500
2020-04-14 11:27:06.000067: I session_manager.py:504] Running local_init_op.
2020-04-14 11:27:06.000375: I session_manager.py:507] Done running local_init_op.
2020-04-14 11:27:08.000858: I train.py:78] Starting additional evaluation
2020-04-14 11:27:08.000859: I train.py:83] Finished additional evaluation
2020-04-14 11:27:18.000519: I train.py:78] Starting additional evaluation
2020-04-14 11:27:18.000520: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000352: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000352: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000412: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000412: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000472: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000473: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000531: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000531: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000588: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000588: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000646: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000646: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000701: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000701: I train.py:83] Finished additional evaluation
2020-04-14 11:27:19.000756: I train.py:78] Starting additional evaluation
2020-04-14 11:27:19.000756: I train.py:83] Finished additional evaluation
2020-04-14 11:27:20.000248: I evaluation.py:273] Inference Time : 18.71248s
2020-04-14 11:27:20.000248: I evaluation.py:276] Finished evaluation at 2020-04-14-11:27:20
```

It runs 10 times, instead of just once. I think it's called for each step, in mine I'm running one epoch for all validation data, but runs in 10 steps (or batches)

**Comment by PeterMcGor** [comment]
Cool! It's true was slow, I didn't have so many datasets so it wast a big problem for me but this it is great! 

Thank you!

**Comment by carranza96** [comment]
Hi. I tried @PeterMcGor and the code inside multi_eval_hook() is executed three times. Any ideas how to solve this? It is actually evaluating the complete set three times, which increases a lot computation time.
